src/config.ts

```diff
@@ -24,16 +27,62 @@ function configOrder(obj: configObjects): configObjects {
   return obj
 }
 
-export function getDefaultShell(): string {
+// finds the default shell start commmand
+async function getDefaultShell(): Promise<string> {
+  let shellStartCommand
+  if (process.platform === "win32") {
+    // Windows
+    try {
+      shellStartCommand = await which("pwsh.exe")
+      return shellStartCommand
+    } catch (e1) {
+      try {
+        shellStartCommand = await which("powershell.exe")
+        return shellStartCommand
+      } catch (e2) {
+        shellStartCommand = process.env.COMSPEC || "cmd.exe"
+        return shellStartCommand
+      }
+    }
+  } else {
+    // Unix
+    shellStartCommand = process.env.SHELL || "/bin/sh"
+    return shellStartCommand
+  }
+}
+
+// The shell command used in case getDefaultShell does not find a prefered shell
+function getFallbackShell() {
   return process.platform === "win32" ? process.env.COMSPEC || "cmd.exe" : process.env.SHELL || "/bin/sh"
 }
 
+// set start command asyncronously
+export async function setShellStartCommand(disposables: CompositeDisposable) {
+  if (atom.config.get("terminal.autoShell")) {
+    const shellStartCommand = await getDefaultShell()
+    atom.config.set("terminal.shell", shellStartCommand)
+  }
+  return disposables.add(
+    // an observre that checks if command has been edited manually, if so it will turn autoShell off
+    atom.config.observe("terminal.shell", () => {
```

Comment:
instead of passing disposables we should return this and add to disposables where setShellStartCommand is called

Reply:
I believe disposables are passed by reference. That's the default behavior of JavaScript for objects/arrays.  That means there is no performance regression.

The downside of returning the disposable is that we need to "await" its promise inside the constructor (which is impossible).

Reply:
I fixed this in https://github.com/atom-ide-community/terminal/pull/24/commits/e1daf500f6d59565724c00827da1025df52cb74a. We don't need to tie adding disposable to setting the default value 😄